pkg/controller/user-cluster-controller-manager/constraint-syncer/controller_test.go

```diff
@@ -0,0 +1,173 @@
+/*
+Copyright 2020 The Kubermatic Kubernetes Platform contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package constraintsyncer
+
+import (
+	"context"
+	"encoding/json"
+	"reflect"
+	"testing"
+	"time"
+
+	kubermaticapiv1 "k8c.io/kubermatic/v2/pkg/api/v1"
+	apiv2 "k8c.io/kubermatic/v2/pkg/api/v2"
+	v1 "k8c.io/kubermatic/v2/pkg/crd/kubermatic/v1"
+	"k8c.io/kubermatic/v2/pkg/handler/test"
+	constrainthandler "k8c.io/kubermatic/v2/pkg/handler/v2/constraint"
+	kubermaticlog "k8c.io/kubermatic/v2/pkg/log"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/diff"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/tools/record"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+	fakectrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client/fake"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	constraintName = "constraint"
+	kind           = "RequiredLabel"
+)
+
+func TestReconcile(t *testing.T) {
+	if err := test.RegisterScheme(test.SchemeBuilder); err != nil {
+		t.Fatal(err)
+	}
+
+	testCases := []struct {
+		name                 string
+		namespacedName       types.NamespacedName
+		expectedConstraint   apiv2.Constraint
+		expectedGetErrStatus metav1.StatusReason
+		seedClient           ctrlruntimeclient.Client
+		userClient           ctrlruntimeclient.Client
+	}{
+		{
+			name: "scenario 1: sync constraint to user cluster",
+			namespacedName: types.NamespacedName{
+				Namespace: "namespace",
+				Name:      constraintName,
+			},
+			expectedConstraint: test.GenDefaultAPIConstraint(constraintName, kind),
+			seedClient: fakectrlruntimeclient.NewFakeClientWithScheme(
+				scheme.Scheme,
+				test.GenConstraint(constraintName, "namespace", kind)),
+			userClient: fakectrlruntimeclient.NewFakeClientWithScheme(scheme.Scheme),
+		},
+		{
+			name: "scenario 3: cleanup ct on user cluster when master ct is being terminated",
```

Comment:
probably this should be `scenario 2`  :-)

Reply:
true true :)